File: src/apps/Andriod/app/src/main/java/com/dkellycollins/triggerfinger/managers/events/impl/EventDispatcher.java
package com.dkellycollins.triggerfinger.managers.events.impl;

import com.dkellycollins.triggerfinger.managers.events.IEventDispatcher;
import com.dkellycollins.triggerfinger.managers.events.IEventInterceptor;
import com.dkellycollins.triggerfinger.managers.events.IEvent;

import java.util.List;

/**
 * Standard implementation of IEventDispatcher
 */
public class EventDispatcher implements IEventDispatcher {

    private final List<IEventInterceptor> _interceptors;

    public EventDispatcher(List<IEventInterceptor> interceptors) {
        _interceptors = interceptors;
    }

    @Override
    public <T extends IEvent> void dispatch(T event) {
        for (IEventInterceptor interceptor :_interceptors) {
            if(!interceptor.handlesMessage(event)) {
                continue;
            }

            interceptor.invoke(event);
        }
    }
}

File: src/apps/Android-Kotlin/app/src/main/kotlin/com/dkellycollins/triggerfinger/managers/events/impl/EventDispatcher.kt
package com.dkellycollins.triggerfinger.managers.events.impl

import com.dkellycollins.triggerfinger.managers.events.IEventDispatcher
import com.dkellycollins.triggerfinger.managers.events.IEventInterceptor
import com.dkellycollins.triggerfinger.managers.events.IEvent

/**
 * Standard implementation of IEventDispatcher
 */
class EventDispatcher(private val _interceptors: List<IEventInterceptor>) : IEventDispatcher {

    override fun <T : IEvent> dispatch(event: T) {
        for (interceptor in _interceptors) {
            if (!interceptor.handlesMessage(event)) {
                continue
            }

            interceptor.invoke(event)
        }
    }
}
